<a href="https://colab.research.google.com/github/rohitLanjewar03/Machine-Learning/blob/main/DL_P1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Implement a multilayer perceptron (MLP) in Python to classify iris flowers into three species based on their sepal length, sepal width, petal length, and petal width.

## Load and explore the data

### Subtask:
Load the iris dataset and explore its structure and content.


**Reasoning**:
The subtask requires loading the iris dataset, exploring its description, and creating a pandas DataFrame to inspect its structure and data types. This can be achieved in a single code block.



In [1]:
from sklearn.datasets import load_iris
import pandas as pd

# Load the iris dataset
iris = load_iris()

# Print the description of the dataset
print(iris.DESCR)

# Create a pandas DataFrame
df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
df['target'] = iris.target

# Display the first few rows and data types
display(df.head())
display(df.info())

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


None

## Preprocess the data

### Subtask:
Split the data into training and testing sets, scale the features, and encode the labels.


**Reasoning**:
Import the necessary libraries for splitting, scaling, and encoding the data.



In [2]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.utils import to_categorical

**Reasoning**:
Separate features and target, split the data into training and testing sets, scale the features, and one-hot encode the labels.



In [3]:
# Separate features (X) and target (y)
X = df.iloc[:, :-1]
y = df['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the scaler to the training data, then transform both training and testing data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Convert target labels to one-hot encoded format
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

## Build the mlp model

### Subtask:
Define the architecture of the multilayer perceptron.


**Reasoning**:
Define the architecture of the multilayer perceptron using Keras Sequential model and Dense layers as per the instructions.



In [4]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Create a Sequential model
model = Sequential()

# Add the input layer
model.add(Dense(64, input_shape=(X.shape[1],), activation='relu'))

# Add a hidden layer
model.add(Dense(32, activation='relu'))

# Add the output layer
model.add(Dense(3, activation='softmax'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


**Reasoning**:
Print a summary of the defined model to verify the architecture and then finish the subtask.



In [5]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 64)             │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 3)              │            99 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,499 (9.76 KB)

 Trainable params: 2,499 (9.76 KB)

 Non-trainable params: 0 (0.00 B)

## Compile and train the model

### Subtask:
Compile and train the model


**Reasoning**:
Compile and train the model using the Adam optimizer, categorical crossentropy loss, and accuracy metric, then train the model for 100 epochs with a batch size of 10.



In [6]:
# Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=100, batch_size=10, verbose=0)

## Evaluate the model

### Subtask:
Evaluate the trained model on the testing data and report the accuracy.


**Reasoning**:
Evaluate the trained model on the testing data and print the accuracy.



In [7]:
# Evaluate the model on the testing data
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)

# Print the test accuracy
print(f'Test accuracy: {accuracy:.4f}')

Test accuracy: 1.0000


## Make predictions

### Subtask:
Make predictions using the trained model.


**Reasoning**:
Use the trained model to predict the classes for the test data and convert the predicted probabilities to class labels.



In [8]:
import numpy as np

# Make predictions on the test data
y_pred = model.predict(X_test)

# Convert the predicted probabilities to class labels
y_pred_classes = np.argmax(y_pred, axis=1)

# Display the predicted classes
print("Predicted classes for the test data:")
print(y_pred_classes)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 236ms/step
Predicted classes for the test data:
[1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0]


## Summary:

### Data Analysis Key Findings

*   The iris dataset, containing 150 instances with 4 numerical features and 3 target classes, was successfully loaded and structured into a pandas DataFrame.
*   The data was split into training and testing sets (80% train, 20% test) with a `random_state` of 42.
*   Feature scaling using `StandardScaler` was applied to both the training and testing sets.
*   Target labels were one-hot encoded for compatibility with the model's output layer.
*   A sequential MLP model was built with an input layer (64 units, ReLU), a hidden layer (32 units, ReLU), and an output layer (3 units, Softmax).
*   The model was compiled using the Adam optimizer, `categorical_crossentropy` loss, and 'accuracy' metric.
*   The model was trained for 100 epochs with a batch size of 10.
*   The trained model achieved a test accuracy of 1.0000.
*   Predictions were made on the test data, and the output probabilities were successfully converted into discrete class labels using `np.argmax`.

### Insights or Next Steps

*   The MLP model achieved perfect accuracy on the test set, suggesting it has learned the patterns in the iris dataset effectively for this specific split.
*   Given the perfect accuracy, further steps could include evaluating the model's robustness using cross-validation or applying it to a new, unseen dataset if available.
